core/src/main/java/org/infinispan/commands/irac/IracPutManyCommand.java

```diff
@@ -0,0 +1,241 @@
+package org.infinispan.commands.irac;
+
+import static org.infinispan.commons.marshall.MarshallUtil.marshallCollection;
+import static org.infinispan.commons.marshall.MarshallUtil.unmarshallCollection;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.commons.util.IntSets;
+import org.infinispan.commons.util.Util;
+import org.infinispan.metadata.Metadata;
+import org.infinispan.metadata.impl.IracMetadata;
+import org.infinispan.util.ByteString;
+import org.infinispan.util.concurrent.AggregateCompletionStage;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.infinispan.xsite.BackupReceiver;
+
+/**
+ * A multi-key cross-site requests.
+ * <p>
+ * This command is used by asynchronous cross-site replication to send multiple keys batched to remote sites. The keys
+ * in this command can include updates, removal or expirations.
+ * <p>
+ * The element order in {@code updateList} is important because the reply will be a {@link IntSet} with the position of
+ * the failed keys.
+ *
+ * @since 14.0
+ */
+public class IracPutManyCommand extends IracUpdateKeyCommand<IntSet> {
+
+   private static final Log log = LogFactory.getLog(IracPutManyCommand.class);
+
+   public static final byte COMMAND_ID = 48;
+
+   private static final byte WRITE = 0;
+   private static final byte REMOVE = 1;
+   private static final byte EXPIRE = 2;
+
+   private List<Update> updateList;
+
+   @SuppressWarnings("unused")
+   public IracPutManyCommand() {
+      super(COMMAND_ID, null);
+   }
+
+   public IracPutManyCommand(ByteString cacheName) {
+      super(COMMAND_ID, cacheName);
+   }
+
+   public IracPutManyCommand(ByteString cacheName, int maxCapacity) {
+      super(COMMAND_ID, cacheName);
+      updateList = new ArrayList<>(maxCapacity);
+   }
+
+   public CompletionStage<IntSet> executeOperation(BackupReceiver receiver) {
+      IntSet rsp = IntSets.mutableEmptySet(updateList.size());
```

Comment:
I would just use a concurrent int set. Then you don't need to synchronize below as well.